bluesky/bundlers.py

```diff
@@ -597,68 +631,274 @@ async def complete(self, msg):
+            if stream_name not in self._descriptor_objs or collect_obj not in self._descriptor_objs[stream_name]:
+                # We do not have an Event Descriptor for this collect_obj.
+                await self._prepare_stream(stream_name, {collect_obj: stream_data_keys})
             else:
                 objs_read = self._descriptor_objs[stream_name]
-                if stream_data_keys != objs_read:
+                if stream_data_keys != objs_read[collect_obj]:
                     raise RuntimeError(
                         "Mismatched objects read, "
                         "expected {!s}, "
                         "got {!s}".format(stream_data_keys, objs_read)
                     )
 
             local_descriptors[frozenset(stream_data_keys)] = self._descriptors[stream_name]
+
         self._local_descriptors[collect_obj] = local_descriptors
 
+    async def _pack_seq_nums_into_stream_datum(
+        self,
+        doc: StreamDatum,
+        message_stream_name: str,
+        stream_datum_previous_indices_difference: int
+    ) -> int:
+
+        if doc["seq_nums"] != StreamRange(start=0, stop=0):
+            raise EventModelValueError(
+                f"Received `seq_nums` {doc['seq_nums']} from devices {doc['data_keys']} "
+                "during `collect()` or `describe_collect()`. `seq_nums` should be None or "
+                "`StreamRange(start=0, stop=0)` on `ComposeStreamDatum` when used with the "
+                "run engine."
+            )
+
+        indices_difference = doc["indices"]["stop"] - doc["indices"]["start"]
+
+        if (
+            stream_datum_previous_indices_difference and
+            stream_datum_previous_indices_difference != indices_difference
+        ):
+            raise EventModelValueError(
+                f"Received `indices` {doc['indices']} during `collect()` these are of a different "
+                f"width `{indices_difference}` than other detectors in the same collect() or save()."
+            )
+
+        current_seq_counter = self._sequence_counters[message_stream_name]
+        doc["seq_nums"] = StreamRange(
+            start=current_seq_counter, stop=current_seq_counter + indices_difference
+        )
+
+        return indices_difference
+
+    async def _pack_external_assets(
+        self,
+        asset_docs: Iterable[Tuple[str, ExternalAssetDoc]],
+        message_stream_name: Optional[str]
+    ):
+        """Packs some external asset documents with relevant information from the run."""
+
+        stream_datum_previous_indices_difference = 0
+        number_of_stream_datum_receieved = 0
+
+        for name, doc in asset_docs:
+            if name == DocumentNames.resource.value:
+                doc["run_start"] = self._run_start_uid
+            elif name == DocumentNames.stream_resource.value:
+                if number_of_stream_datum_receieved:
+                    raise RuntimeError("Received a `stream_resource` document after `stream_datum` documents.")
+
```

Comment:
This is valid for the case of a flyer that wraps 2 detectors

Reply:
Instead we should cache in self the stream_resource by uid, then check when we get a stream_datum that it refers to a stream_resource that exists